pkg/k8s/cnpstatus.go

```diff
@@ -0,0 +1,301 @@
+				// controller) before consuming the next event.
+				ch <- nsu
+			}
+		}
+	}
+}
+
+// StopController stops the controller which is managing the sending of
+// status updates to the Kubernetes APIServer for the given CNP.
+func (c *CNPStatusEventHandler) StopController(cnp *types.SlimCNP) {
+	cnpKey := getKeyFromObjectMeta(cnp.ObjectMeta)
+	prefix := path.Join(CNPStatusesPath, cnpKey)
+	err := kvstore.DeletePrefix(prefix)
+	if err != nil {
+		log.WithError(err).WithField("prefix", prefix).Warning("error deleting prefix from kvstore")
+	}
+	err = c.controllers.RemoveController(cnpKey)
+	if err != nil {
+		log.WithError(err).WithField(logfields.Controller, cnpKey).Warning("error removing controller")
+	}
+}
+
+// StartController starts the controller which sends status updates for the
+// given CNP to the Kubernetes APIserver.
+func (c *CNPStatusEventHandler) StartController(cnp *types.SlimCNP) {
+	cnpKey := path.Join(string(cnp.UID), cnp.Namespace, cnp.Name)
+	ch, ok := c.eventMap.createIfNotExist(cnpKey)
+	if ok {
+		return
+	}
+
+	namespace := cnp.Namespace
+	name := cnp.Name
+
+	nodeStatusMap := make(map[string]cilium_v2.CiliumNetworkPolicyNodeStatus)
+	var hasRun bool
+	c.controllers.UpdateController(cnpKey, controller.ControllerParams{
+		DoFunc: func(ctx context.Context) error {
+			// Get any updates received in case other nodes updated the kvstore
+			// before this controller was started, as cilium-agents running in
+			// the cluster may receive the CNP event from their own Kubernetes
+			// watchers and send keys for their status to the key-value store
+			// before this controller has been started.
+			if !hasRun {
+				sharedKeys := c.cnpStore.SharedKeysMap()
+				for k := range sharedKeys {
+					// Look for any key which matches this CNP.
+					if strings.HasPrefix(k, cnpKey) {
+						v, ok := sharedKeys[k].(*CNPNSWithMeta)
+						if !ok {
+							log.Errorf("received unexpected type mapping to key %s in cnp shared store: %T", k, sharedKeys[k])
+							continue
+						}
+						nodeStatusMap[v.Node] = v.CiliumNetworkPolicyNodeStatus
+					}
+				}
+				hasRun = true
+			} else {
+				// Wait for one status update to occur.
+				select {
+				case ev, ok := <-ch:
+					if ok {
+						nodeStatusMap[ev.node] = *ev.CiliumNetworkPolicyNodeStatus
+					}
+				case <-ctx.Done():
+					return nil
+				}
+			}
+
+		Loop:
+			for {
+				select {
+				case <-ctx.Done():
+					// Controller was stopped, we can simply exit.
+					return nil
+				case ev, ok := <-ch:
+					if ok {
+						nodeStatusMap[ev.node] = *ev.CiliumNetworkPolicyNodeStatus
+					}
+				default:
+					// Nothing to do, update K8s apiserver with status updates.
+					break Loop
+				}
+			}
+
+			// This will only be the case after the first run has occurred and
+			// there are no subsequent status updates since we don't want for
+			// at least one event in the first run.
+			if len(nodeStatusMap) == 0 {
+				return nil
+			}
```

Comment:
I don't understand this. This seems to be solving something that can be avoided with a rewrite of this function's logic.

Reply:
My goal here was to ensure that we get all updates that came before we started this goroutine from the SharedStore, but that if we have received updates from the shared store, to not block on receiving at least one node status update from the channel, because we already "received" them from the store; during the first run, there are two sources to 'receive' updates - the map is the 'receiver' of these updates, so checking whether the map has been updated seems like a reasonable check to me here. If more updates are sent after we start the goroutine (by reading from the channel on line 235), we will batch those and send them to the K8s apiserver all in one go. In all other cases, except for the first run, we wait for at least one event to be received, and then wait for subsequent events or until no more subsequent events are received at some time.

Reply:
Next push will have this explained more clearly / performed more cleanly.